stock_location_orderpoint/models/stock_location_orderpoint.py

```diff
@@ -0,0 +1,429 @@
+        virtual_available_on_dest = qties_on_dest["virtual_available"]
+        if (
+            float_compare(
+                virtual_available_on_dest, 0, precision_rounding=product.uom_id.rounding
+            )
+            >= 0
+        ):
+            return 0
+
+        virtual_available_on_dest = abs(virtual_available_on_dest)
+        qties_on_src = qties_on_locations[self.location_src_id][product]
+        virtual_available_on_src = (
+            qties_on_src["virtual_available"] - qties_on_src["incoming_qty"]
+        )
+        if (
+            float_compare(
+                virtual_available_on_src,
+                0,
+                precision_rounding=product.uom_id.rounding,
+            )
+            <= 0
+        ):
+            return 0
+
+        qty_to_replenish = virtual_available_on_dest - qty_already_replenished
+        return min(qty_to_replenish, virtual_available_on_src)
+
+    def _get_qties_to_replenish(self, moves_by_location):
+        products = self.env["product.product"]
+        for moves in moves_by_location.values():
+            products |= moves.product_id
+        qties_on_locations = self._compute_quantities_dict(
+            (self.location_id | self.location_src_id), products
+        )
+        qties_replenished = {}
+        qties_to_replenish = {}
+        for orderpoint in self:
+            if orderpoint.location_id not in moves_by_location:
+                continue
+
+            for product in moves_by_location[orderpoint.location_id].product_id:
+                qties_replenished_for_location = qties_replenished.setdefault(
+                    orderpoint.location_id, {}
+                )
+                qties_replenished_for_location.setdefault(product, 0)
+
+                qty_to_replenish = orderpoint._get_qty_to_replenish(
+                    product,
+                    qties_on_locations,
+                    qties_replenished_for_location[product],
+                )
+                if (
+                    float_compare(
+                        qty_to_replenish, 0, precision_rounding=product.uom_id.rounding
+                    )
+                    > 0
+                ):
+                    qties_to_replenish.setdefault(
+                        orderpoint,
+                        [],
+                    )
+                    qties_to_replenish[orderpoint].append((product, qty_to_replenish))
+                    qties_replenished_for_location[product] += qty_to_replenish
+        return qties_to_replenish
+
+    def __prepare_procurements(self, moves_by_location):
+        qties_to_replenish_by_orderpoint = self._get_qties_to_replenish(
+            moves_by_location
+        )
+        procurements = []
+        for orderpoint, qties_to_replenish in qties_to_replenish_by_orderpoint.items():
+            proc_vals = orderpoint._prepare_procurement_values()
+            for product, qty in qties_to_replenish:
+                date_planned = min(
```

Comment:
Could you move that part of code in separate method ?

I've tested that in stock.move with success:

```python
    @ormcache("moves", "product")
    def _get_location_orderpoint_replenishment_date(self, moves, product):
        return min(moves.filtered(
            lambda move: move.product_id == product).mapped("date"))
```

Then:

```suggestion
                date_planned = date_planned = move_obj._get_location_orderpoint_replenishment_date(
                    moves_by_location[orderpoint.location_id], product)
```